### PR TITLE
[java] JUnitTestsShouldIncludeAssertRule supports @Rule annotated ExpectedExceptions

### PR DESCRIPTION
**Pull request by Oowekyala**
**Fixes:** #285 
**PR Description:** JUnitTestsShouldIncludeAssertRule no longer reports methods which are not annotated with `@Test(expected=<exceptionclass>)` nor contain assert statements, provided the test class has a field of type `ExpectedException` annotated with `@Rule`, and the Test calls at least one method starting by `expect` on that field.

See #285 for more details.

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java [hunk 1]
```diff
@@ -29,36 +37,73 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
...
+     */
+    private Map<String, List<NameOccurrence>> getRuleAnnotatedExpectedExceptions(Scope classScope) {
+        Map<String, List<NameOccurrence>> result = new HashMap<>();
+        Map<NameDeclaration, List<NameOccurrence>> decls = classScope.getDeclarations();
+
+        for (NameDeclaration decl : decls.keySet()) {
+            Node parent = decl.getNode().jjtGetParent().jjtGetParent().jjtGetParent();
+            if (parent.hasDescendantOfType(ASTAnnotation.class)
+                    && parent.getFirstChildOfType(ASTFieldDeclaration.class) != null) {
+                if (!"Rule"
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java]: bare in mind that, even if non-conventional, people may simply write `@org.junit.Rule`too

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java [hunk 2]
```diff
@@ -29,36 +37,73 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
...
+                        .equals(parent.getFirstDescendantOfType(ASTMarkerAnnotation.class).jjtGetChild(0).getImage())) {
+                    System.out.println(
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java]: avoid these outputs

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java [hunk 3]
```diff
@@ -29,36 +37,73 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
...
+                            parent.getFirstDescendantOfType(ASTMarkerAnnotation.class).jjtGetChild(0).getImage());
+                    continue;
+                }
 
+                Node type = parent.getFirstDescendantOfType(ASTReferenceType.class);
+                if (!"ExpectedException".equals(type.jjtGetChild(0).getImage())) {
+                    System.out.println(type.jjtGetChild(0).getImage());
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java]: ditto

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java (hunk 1 above)
Comment (Oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/junit/JUnitTestsShouldIncludeAssertRule.java]: You're right, I added support for that!

**Comment by jsotuyod** [comment]
Merged! Awesome PR, this will be part of the next release cycle!

**Pull request merged**